### ISSUE TITLE
Database scan/compaction at startup takes too long

### ISSUE THREAD

**Issue by sgtnoodle** [syncthing/syncthing#2439]
I'm running "v0.12.1, Linux (ARM)" in arch linux on a raspberry pi. After starting up syncthing, it takes several minutes before the thing actually starts listening on port 8384.

I would expect syncthing to open up its listening sockets within a couple seconds of starting up.

```
[xxx@xxx~]$ syncthing -verbose
[monitor] 10:53:21 INFO: Starting syncthing
[C63OC] 10:53:21 INFO: syncthing v0.12.1 "Beryllium Bedbug" (go1.5.1 linux-arm default) builduser@leming 2015-11-06 16:14:20 UTC
[C63OC] 10:53:21 INFO: My ID: XXX
[C63OC] 10:53:21 INFO: Single thread hash performance is ~3.9 MB/s
[C63OC] 10:53:21 VERBOSE: Starting up (/home/xxx/.config/syncthing)
[C63OC] 10:53:21 OK: Ready to synchronize default (read-write)
[C63OC] 10:53:21 INFO: Using discovery server https://discovery-v4-1.syncthing.net/?id=XXX
[C63OC] 10:53:21 INFO: Using discovery server https://discovery-v4-2.syncthing.net/?id=XXX
[C63OC] 10:53:21 INFO: Using discovery server https://discovery-v4-3.syncthing.net/?id=XXX
[C63OC] 10:53:21 INFO: Using discovery server https://discovery-v6-1.syncthing.net/?id=XXX
[C63OC] 10:53:21 INFO: Using discovery server https://discovery-v6-2.syncthing.net/?id=XXX
[C63OC] 10:53:21 INFO: Using discovery server https://discovery-v6-3.syncthing.net/?id=XXX
[C63OC] 10:53:21 INFO: Loading HTTPS certificate: open /home/xxx/.config/syncthing/https-cert.pem: no such file or directory
[C63OC] 10:53:21 INFO: Creating new HTTPS certificate
[C63OC] 10:53:21 VERBOSE: Folder "default" is now scanning
[C63OC] 10:53:21 INFO: Completed initial scan (rw) of folder default
[C63OC] 10:53:21 VERBOSE: Folder "default" is now idle
[C63OC] 10:53:31 VERBOSE: Discovered device XXX at [{tcp://0.0.0.0:22000}]
[C63OC] 10:53:31 VERBOSE: Discovered device XXX at [{tcp://0.0.0.0:22000}]
[C63OC] 10:53:31 VERBOSE: Relay state changed; connected relay(s) are relay://xxx:22067/?id=XXX&pingInterval=1m0s&networkTimeout=2m0s&sessionLimitBps=1000000&globalLimitBps=2000000&statusAddr=:22070&providedBy=thirtythreeforty.
[C63OC] 10:53:31 INFO: No UPnP device detected
[C63OC] 10:54:31 VERBOSE: Folder "default" is now scanning
[C63OC] 10:54:31 VERBOSE: Folder "default" is now idle
[C63OC] 10:55:10 INFO: Device XXX is "xxx" at [dynamic]
[C63OC] 10:55:10 VERBOSE: Startup complete
[C63OC] 10:55:11 INFO: API listening on 127.0.0.1:8384
[C63OC] 10:55:11 INFO: GUI URL is http://127.0.0.1:8384/
[C63OC] 10:55:29 VERBOSE: Folder "default" is now scanning
[C63OC] 10:55:29 VERBOSE: Folder "default" is now idle
```

edit: Use code blocks.

**Comment by calmh** [comment]
Looks like this was the first startup and your computer is rather slow. 

INFO: Creating new HTTPS certificate

That took a while. Next startup should be fine.

**Comment by sgtnoodle** [comment]
Thanks for the quick response! I apologize for re-opening the issue, but I don't think that's the root cause. It's true that I used a new user to try to isolate the problem with all default settings. On my real user, the wait is 30 minutes.

 Nov 07 11:02:51 xxx systemd[1]: Started Syncthing - Open Source Continuous File Synchronization for syncthing.
Nov 07 11:02:51 xxx syncthing[592]: [O7BJA] INFO: syncthing v0.12.1 "Beryllium Bedbug" (go1.5.1 linux-arm default) builduser@leming 2015-11-06 16:14:20 UTC
Nov 07 11:02:51 xxx syncthing[592]: [O7BJA] INFO: My ID: XXX
Nov 07 11:02:52 xxx syncthing[592]: [O7BJA] INFO: Single thread hash performance is ~3.9 MB/s
Nov 07 11:35:28 xxx syncthing[592]: [O7BJA] OK: Ready to synchronize storage (read-write)
Nov 07 11:35:28 xxx syncthing[592]: [O7BJA] INFO: Using discovery server https://discovery-v4-1.syncthing.net/?id=XXX
Nov 07 11:35:28 xxx syncthing[592]: [O7BJA] INFO: Using discovery server https://discovery-v4-2.syncthing.net/?id=XXX
Nov 07 11:35:28 xxx syncthing[592]: [O7BJA] INFO: Using discovery server https://discovery-v4-3.syncthing.net/?id=XXX
Nov 07 11:35:28 xxx syncthing[592]: [O7BJA] INFO: Using discovery server https://discovery-v6-1.syncthing.net/?id=XXX
Nov 07 11:35:28 xxx syncthing[592]: [O7BJA] INFO: Using discovery server https://discovery-v6-2.syncthing.net/?id=XXX
Nov 07 11:35:28 xxx syncthing[592]: [O7BJA] INFO: Using discovery server https://discovery-v6-3.syncthing.net/?id=XXX
Nov 07 11:35:28 xxx syncthing[592]: [O7BJA] INFO: Device XXX is "xxx" at [dynamic]
Nov 07 11:35:28 xxx syncthing[592]: [O7BJA] INFO: Device XXX is "xxx" at [dynamic]
Nov 07 11:35:31 xxx syncthing[592]: [O7BJA] INFO: API listening on 127.0.0.1:8384
Nov 07 11:35:31 xxx syncthing[592]: [O7BJA] INFO: GUI URL is http://127.0.0.1:8384

This line in particular seems relevant:
Nov 07 11:35:28 xxx syncthing[592]: [O7BJA] OK: Ready to synchronize storage (read-write)

Even with a slow computer, it was rather non-intuitive to have a non-working admin interface and no log message saying something to the effect, "This operation may take a long time!" Perhaps the admin interface could be fired up earlier in the startup process, or at the least a comforting log message added?

**Comment by calmh** [comment]
Mmmkay... So what happens between "My ID: ..." and "Ready to synchronize" is mostly that we open the database, which results in it being scanned for some incorrect data and to build up the in memory size info. It seems that takes ages for you. How much data is there in the "storage" folder, and how many files?

**Comment by alex2108** [comment]
sounds like the performance of my raspberry pi, but it's less confusing for me because a lot of those `Ready to synchronize` messages appear because my data is distributed between multiple folders, but yeah, it needs long... (100% CPU usage during that time)

```
   "totFiles" : 67440,
   "totMiB" : 62614,
```

```
[NRL2J] 2015/11/08 20:10:20.225366 main.go:515: INFO: syncthing v0.12.1 "Beryllium Bedbug" 
[...] a lot of "Ready to synchronize" messages and other stuff
[NRL2J] 2015/11/08 20:38:17.240661 gui.go:169: INFO: API listening on [::]:8384
[NRL2J] 2015/11/08 20:38:17.241188 gui.go:169: INFO: GUI URL is http://127.0.0.1:8384/
```

**Comment by calmh** [comment]
That's... Wow. It feels like I could size and verify 67 thousand files by hand faster than what happens there. :/

**Comment by sgtnoodle** [comment]
I have ~650000 files / ~1TB of data on a windows desktop. The raspberry pi isn't yet fully synced up, and at the time, probably had about 60,000 files. This is a science experiment. :-)

Interestingly, the windows box wanted to upgrade itself to 0.12.2 (it was complaining about syncthing.exe.old already existing) so I killed it, deleted the .old file and restarted it. It came back up with 0.12.1 and the raspberry pi re-connected. A few seconds later the windows box updated itself to 0.12.2 and restarted, and now the raspberry pi isn't re-connecting. I connected to the pi's web gui and it says its local state is 0 items / ~0 B. I don't know if syncthing is backwards-compatible across versions or not, so I'll go ahead and try to update the raspberry pi. I reckon it should take a couple hours to bring up its sockets this time, since it's been syncing for a few more days!

**Comment by sgtnoodle** [comment]
Is this a duplicate of #2458 ?

**Comment by AudriusButkevicius** [comment]
Simillar, but not the same.

**Comment by martin-rueegg** [comment]
the same here. what makes it a bit frustrating, is the fact that synchronization already starts before the GUI is available. so one is blind during the first "minutes" (in my case: 2') ...

```
[monitor] 01:28:17 INFO: Starting syncthing
[AGWY5] 01:28:17 INFO: syncthing v0.12.19 "Beryllium Bedbug" (go1.5.3 linux-amd64) unknown-user@build2.syncthing.net 2016-02-14 19:26:33 UTC
[AGWY5] 01:28:17 INFO: My ID: XXXXXXX-XXXXXXX-XXXXXXX-XXXXXXX-XXXXXXX-XXXXXXX-XXXXXXX-AAAAAAA
[AGWY5] 01:28:17 INFO: Single thread hash performance is ~208 MB/s
[AGWY5] 01:28:19 OK: Ready to synchronize foo (read-write)
[AGWY5] 01:30:19 OK: Ready to synchronize bar (read-write)
[AGWY5] 01:30:19 OK: Ready to synchronize default (read-write)
[AGWY5] 01:30:21 OK: Ready to synchronize adfadf (read-write)
[AGWY5] 01:30:21 OK: Ready to synchronize aqre (read-write)
[AGWY5] 01:30:21 INFO: Using discovery server https://my.host.name:8443
[AGWY5] 01:30:21 INFO: Completed initial scan (rw) of folder default
[AGWY5] 01:30:21 INFO: Device XXXXXXX-XXXXXXX-XXXXXXX-XXXXXXX-XXXXXXX-XXXXXXX-XXXXXXX-AAAAAAA is "aaaa" at [dynamic]
[AGWY5] 01:30:21 INFO: Device XXXXXXX-XXXXXXX-XXXXXXX-XXXXXXX-XXXXXXX-XXXXXXX-XXXXXXX-BBBBBBB is "bbbb" at [dynamic]
[AGWY5] 01:30:21 INFO: Device XXXXXXX-XXXXXXX-XXXXXXX-XXXXXXX-XXXXXXX-XXXXXXX-XXXXXXX-CCCCCCC is "cccc" at [dynamic]
[AGWY5] 01:30:21 INFO: Starting usage reporting
[AGWY5] 01:30:21 INFO: API listening on [::]:8384
[AGWY5] 01:30:21 INFO: GUI URL is https://127.0.0.1:8384/
...
[AGWY5] 01:30:22 INFO: Completed initial scan (rw) of folder aqre
[AGWY5] 01:30:22 INFO: Completed initial scan (rw) of folder adfadf
[AGWY5] 01:30:24 INFO: Completed initial scan (rw) of folder bar
[AGWY5] 01:30:31 INFO: New UPnP port mapping: external port 36824 to local port 22000.
[AGWY5] 01:38:20 INFO: Completed initial scan (rw) of folder foo
```

this is also a problem, if i need to stop a certain device from sync. (i know, i can start with `-paused`) but that's not practical when installed as a service and doen't apply for the restart action, in which case the GUI remains "unresponsive" for the same amount of time, of course.

**Comment by AudriusButkevicius** [comment]
It's all relative to the device it's running on.

**Comment by martin-rueegg** [comment]
i think the order in which program routines are called do not necessarily depend on the device.

the gui could (in theory) be made available as soon as the program starts, showing limited information at first and then more details as soon as the other "components" come life. the list of folders and devices could be presented after reading the `config.xml`

(by the way, this happens on two different machines, quite decent ones, with plenty of RAM (>=32GB) and of course multple cores. however the folders have quite huge number of files. that's probably why.)

nevertheless i think it makes sense to re-enigneer the start-up sequence to make the GUI available before accepting sync-connections. and rather show a "please wait ..." message, than not having the port listening at all.

**Comment by canton7** [comment]
Please no. The GUI used to start earlier, and there was significant pain and complexity associated with trying to load a bunch of startup data from SyncTrayzor (and I'm sure the other wrappers experienced the same). 

I guess it's one of those things where if you're the one prepared to make the (pretty significant) effort and contribute this change (and also hopefully fix the various wrappers :wink:) then it will be merged. Otherwise the developers have bigger fish to fry.

**Comment by martin-rueegg** [comment]
i see. well, it's my personal opinion that a software _having_ a GUI, should show the GUI at the earliest possible moment, so that the user is informed about the progress. even a "Please wait ..." page (maybe showing the log output) that would load once the software is up, would be great, as now the browser just can't connect and will not retry by itself.

_And_ I also think that a software should not start major background activities (like synch'ing, in this case) unless the user can interact with it and possibly pause a device or shut down the service.

but of course, that is my personal opinion. and of course a kill can always be done on the terminal, after fiddling out the PID (as i don't want to killall those synchting processes at once - well, i have a init.d wrapper script that saves the PID and can stop it quickly ..).

maybe a won't-fix would then help to clarify the position.

**Comment by PythonNut** [comment]
This is also relevant on Syncthing for Android, which displays a "Loading..." dialog while it waits for the socket to become open. Android devices are still pretty underpowered for this sort of thing, and large folders can take quite a while to index (on the order of an hour). 

It would be nice to show the UI immediately, both so the user can see the progress of the initial scan, and possibly execute other actions while the scan is taking place. (Adding and removing devices, changing settings etc.)

**Comment by calmh** [comment]
We don't do database compaction at startup any more, so some of this has been reduced. We do however do a full database scan as this is necessary to build up the caches for how much data we have, etc. This is essentially the same amount of work needed to a single folder scan, so it's not unreasonable. However at startup time all data is uncached and on slow devices it does take a while.

We could spin up the API/GUI earlier, but it won't be able to answer many of the questions asked of it until this is done, so we need to return something like "503 service unavailable" in the meantime. The GUI then needs to handle that which makes everything more complex, and the end result is anyway they GUI showing something like "Loading..." because there's not much else it can do.

**Comment by martin-rueegg** [comment]
well, as pointed out in my [previous comment](https://github.com/syncthing/syncthing/issues/2439#issuecomment-193172267), i think being able to terminate a software _would_ indeed be beneficial, especially for low-resource devices.

the android app does indeed just show waiting, which can take - as pointed out by @PythonNut take up to an hour (on my device as well). yes, also on android there are ways to kill an unresponsive service, however doing it from a GUI would be a) nicer and b) available to the lay person as well. furthermore the information, that syncthing is _scanning_ would be appreciated, also to see the resource consumption.

I do appreciate that the implementation of partial information would require quite some work. however in my opinion, this is a design error and the fact that it hasn't been considered earlier should not be an argument to keep it like that. Without knowing the code, it is also not clear to me, what functions exactly are not available during the initial scan but during later scans. 

- showing the configuration
- and resource usage,
- adding new folders
- or devices, 
- maybe removing folders, 
- pausing the sync (synching is started before the GUI is present!) with a specific remote device and 
- many other things

should (theoretically) still be possible.

however, this is basically not the topic of this ticket. :-)

**Comment by canton7** [comment]
> I do appreciate that the implementation of partial information would require quite some work

If only you knew the pain that starting the API before it was able to function has caused people like me in the past...

**Comment by calmh** [comment]
Unfortunately it takes as long as it needs to take...
